File: app/src/main/java/com/rowland/delivery/features/dash/presentation/adapters/OrderItemAdapter.java
package com.rowland.delivery.features.dash.presentation.adapters;

import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.rowland.delivery.features.dash.domain.models.order.OrderItem;
import com.rowland.delivery.features.dash.presentation.tools.recylerview.HFRecyclerView;
import com.rowland.delivery.merchant.R;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Rowland on 5/9/2018.
 */

public class OrderItemAdapter extends HFRecyclerView<OrderItem> {

    public OrderItemAdapter() {
        super(null, false, false);
    }

    public OrderItemAdapter(boolean withHeader, boolean withFooter) {
        super(null, withHeader, withFooter);
    }

    public OrderItemAdapter(List<OrderItem> data, boolean withHeader, boolean withFooter) {
        super(data, withHeader, withFooter);
    }

    @Override
    protected RecyclerView.ViewHolder getItemView(LayoutInflater inflater, ViewGroup parent) {
        return new OrderItemViewHolder(inflater.inflate(R.layout.row_single_order_item, parent, false));
    }

    @Override
    protected RecyclerView.ViewHolder getHeaderView(LayoutInflater inflater, ViewGroup parent) {
        return null;
    }

    @Override
    protected RecyclerView.ViewHolder getFooterView(LayoutInflater inflater, ViewGroup parent) {
        return null;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof OrderItemViewHolder) {
            OrderItemViewHolder itemHolder = (OrderItemViewHolder) holder;
            OrderItem data = getItem(position);
            itemHolder.bind(data);
        } else if (holder instanceof HeaderViewHolder) {

        } else if (holder instanceof FooterViewHolder) {

        }
    }

    public void setList(final List<OrderItem> orders) {
        if (this.data == null) {
            this.data = orders;
            notifyItemRangeInserted(0, orders.size());
        } else {
            DiffUtil.DiffResult result = DiffUtil.calculateDiff(new DiffUtil.Callback() {
                @Override
                public int getOldListSize() {
                    return OrderItemAdapter.this.data.size();
                }

                @Override
                public int getNewListSize() {
                    return orders.size();
                }

                @Override
                public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                    return OrderItemAdapter.this.data.get(oldItemPosition).getItemCode() ==
                            orders.get(newItemPosition).getItemCode();
                }

                @Override
                public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                    OrderItem newItem = orders.get(newItemPosition);
                    OrderItem oldItem = orders.get(oldItemPosition);
                    return newItem.getItemCode() == oldItem.getItemCode()
                            && newItem.equals(oldItem);
                }
            });
            this.data = orders;
            result.dispatchUpdatesTo(this);
        }
    }

    public static class OrderItemViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.food_image)
        ImageView food_image;

        @BindView(R.id.items_count)
        TextView items_count;

        @BindView(R.id.item_name)
        TextView item_name;

        @BindView(R.id.item_price)
        TextView item_price;

        @BindView(R.id.item_total_price)
        TextView item_total_price;


        public OrderItemViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        public void bind(OrderItem orderItem) {
            Glide.with(food_image.getContext()).load(orderItem.getImageUrl()).into(food_image);
            items_count.setText(String.valueOf(orderItem.getItemQuantity()));
            item_name.setText(orderItem.getTitle());
            item_price.setText("Price: Ksh " + orderItem.getPrice() + " each");
            item_total_price.setText("Total price: Ksh " + orderItem.getItemTotal());
        }
    }

    public static class HeaderViewHolder extends RecyclerView.ViewHolder {
        public HeaderViewHolder(View itemView) {
            super(itemView);
        }
    }

    public static class FooterViewHolder extends RecyclerView.ViewHolder {
        public FooterViewHolder(View itemView) {
            super(itemView);
        }
    }
}

File: app/src/main/java/com/rowland/delivery/features/dash/presentation/adapters/OrderItemAdapter.kt
package com.rowland.delivery.features.dash.presentation.adapters

import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import butterknife.ButterKnife
import com.bumptech.glide.Glide
import com.rowland.delivery.features.dash.domain.models.order.OrderItem
import com.rowland.delivery.features.dash.presentation.tools.recylerview.HFRecyclerView
import com.rowland.delivery.merchant.R
import kotlinx.android.synthetic.main.row_single_order_item.view.*


/**
 * Created by Rowland on 5/9/2018.
 */

class OrderItemAdapter : HFRecyclerView<OrderItem> {

    constructor() : super(null, false, false) {}

    constructor(withHeader: Boolean, withFooter: Boolean) : super(null, withHeader, withFooter) {}

    constructor(data: List<OrderItem>, withHeader: Boolean, withFooter: Boolean) : super(data, withHeader, withFooter) {}

    override fun getItemView(inflater: LayoutInflater, parent: ViewGroup): androidx.recyclerview.widget.RecyclerView.ViewHolder {
        return OrderItemViewHolder(inflater.inflate(R.layout.row_single_order_item, parent, false))
    }

    override fun getHeaderView(inflater: LayoutInflater, parent: ViewGroup): androidx.recyclerview.widget.RecyclerView.ViewHolder? {
        return null
    }

    override fun getFooterView(inflater: LayoutInflater, parent: ViewGroup): androidx.recyclerview.widget.RecyclerView.ViewHolder? {
        return null
    }

    override fun onBindViewHolder(holder: androidx.recyclerview.widget.RecyclerView.ViewHolder, position: Int) {
        if (holder is OrderItemViewHolder) {
            val data = getItem(position)
            holder.bind(data)
        } else if (holder is HeaderViewHolder) {

        } else if (holder is FooterViewHolder) {

        }
    }

    fun setList(orders: List<OrderItem>) {
        if (this.data == null) {
            this.data = orders
            notifyItemRangeInserted(0, orders.size)
        } else {
            val result = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize(): Int {
                    return this@OrderItemAdapter.data.size
                }

                override fun getNewListSize(): Int {
                    return orders.size
                }

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return this@OrderItemAdapter.data[oldItemPosition].itemCode === orders[newItemPosition].itemCode
                }

                override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    val newItem = orders[newItemPosition]
                    val oldItem = orders[oldItemPosition]
                    return newItem.itemCode === oldItem.itemCode && newItem == oldItem
                }
            })
            this.data = orders
            result.dispatchUpdatesTo(this)
        }
    }

    class OrderItemViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {

        init {
            ButterKnife.bind(this, itemView)
        }

        fun bind(orderItem: OrderItem) {
            Glide.with(itemView.food_image!!.context).load(orderItem.imageUrl).into(itemView.food_image!!)
            itemView.items_count!!.text = orderItem.itemQuantity.toString()
            itemView.item_name!!.text = orderItem.title
            itemView.item_price!!.text = "Price: Ksh " + orderItem.price + " each"
            itemView.item_total_price!!.text = "Total price: Ksh " + orderItem.itemTotal!!
        }
    }

    class HeaderViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView)

    class FooterViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView)
}
